File: app/models/api_key.py
"""API Key 数据模型

支持 API Key 认证，用于服务间调用和 MCP 服务器访问。
"""

from datetime import UTC, datetime
from enum import Enum

from sqlalchemy import DateTime, Index, Text
from sqlalchemy.dialects.postgresql import ARRAY, JSONB
from sqlmodel import Column, Field, SQLModel

# BaseModel should be SQLModel from sqlmodel, not from database
BaseModel = SQLModel


class ApiKeyStatus(str, Enum):
    """API Key 状态"""

    ACTIVE = "active"
    INACTIVE = "inactive"
    REVOKED = "revoked"
    EXPIRED = "expired"


class ApiKeyType(str, Enum):
    """API Key 类型"""

    PERSONAL = "personal"  # 个人 API Key
    SERVICE = "service"  # 服务间调用
    MCP = "mcp"  # MCP 服务器专用
    WEBHOOK = "webhook"  # Webhook 验证


class ApiKeyBase(SQLModel):
    """API Key 基础模型"""

    name: str = Field(max_length=100, description="API Key 名称")
    key_prefix: str = Field(max_length=20, description="API Key 前缀（用于显示）")
    key_type: ApiKeyType = Field(default=ApiKeyType.PERSONAL, description="API Key 类型")
    status: ApiKeyStatus = Field(default=ApiKeyStatus.ACTIVE, description="API Key 状态")
    user_id: int = Field(index=True, description="关联的用户 ID")
    scopes: list[str] = Field(default_factory=list, sa_column=Column(ARRAY(Text)), description="权限范围")
    expires_at: datetime | None = Field(default=None, description="过期时间")
    last_used_at: datetime | None = Field(default=None, description="最后使用时间")
    rate_limit: int | None = Field(default=None, description="速率限制（每分钟请求数）")


class ApiKey(ApiKeyBase, BaseModel, table=True):
    """API Key 表模型"""

    __tablename__ = "api_keys"

    id: int | None = Field(default=None, primary_key=True)

    # 存储加密后的完整 Key
    hashed_key: str = Field(max_length=255, description="加密后的 API Key")

    # 元数据
    description: str | None = Field(default=None, sa_column=Column(Text), description="描述")
    extra_data: dict | None = Field(default=None, sa_column=Column(JSONB), description="扩展元数据")

    # 索引
    __table_args__ = (
        Index("ix_api_keys_user_status", "user_id", "status"),
        Index("ix_api_keys_key_type", "key_type"),
        Index("ix_api_keys_expires_at", "expires_at"),
    )


class ApiKeyCreate(SQLModel):
    """创建 API Key 请求模型"""

    name: str = Field(..., max_length=100, description="API Key 名称")
    key_type: ApiKeyType = Field(default=ApiKeyType.PERSONAL, description="API Key 类型")
    scopes: list[str] = Field(default_factory=list, description="权限范围")
    expires_in_days: int | None = Field(default=None, description="有效期（天数）")
    description: str | None = Field(default=None, description="描述")
    rate_limit: int | None = Field(default=None, description="速率限制（每分钟请求数）")


class ApiKeyUpdate(SQLModel):
    """更新 API Key 请求模型"""

    name: str | None = Field(default=None, max_length=100)
    status: ApiKeyStatus | None = Field(default=None)
    scopes: list[str] | None = Field(default=None)
    expires_at: datetime | None = Field(default=None)


class ApiKeyRead(SQLModel):
    """API Key 读取模型"""

    id: int
    name: str
    key_prefix: str
    key_type: ApiKeyType
    status: ApiKeyStatus
    scopes: list[str]
    expires_at: datetime | None
    last_used_at: datetime | None
    created_at: datetime
    updated_at: datetime


class ApiKeyResponse(SQLModel):
    """创建 API Key 响应模型（仅返回一次完整 Key）"""

    id: int
    name: str
    key: str  # 完整的 API Key（仅创建时返回）
    key_prefix: str
    key_type: ApiKeyType
    status: ApiKeyStatus
    scopes: list[str]
    expires_at: datetime | None
    created_at: datetime


class ApiKeyVerifyResponse(SQLModel):
    """API Key 验证响应"""

    valid: bool
    api_key_id: int | None
    user_id: int | None
    scopes: list[str]
    key_type: ApiKeyType

File: frontend/src/types/apiKey.ts
/**
 * Kiki Agent Framework - API Key 相关类型
 */

/**
 * API Key 类型
 */
export type ApiKeyType = 'personal' | 'service' | 'mcp' | 'webhook';

/**
 * API Key 状态
 */
export type ApiKeyStatus = 'active' | 'revoked' | 'expired';

/**
 * API Key 信息（列表用，不含完整 key）
 */
export interface ApiKey {
  id: number;
  name: string;
  key_prefix: string;
  key_type: ApiKeyType;
  status: ApiKeyStatus;
  scopes: string[];
  expires_at: string | null;
  created_at: string;
  last_used: string | null;
  description?: string;
  rate_limit?: number;
}

/**
 * 创建 API Key 请求
 */
export interface CreateApiKeyRequest {
  name: string;
  key_type: ApiKeyType;
  scopes: string[];
  expires_in_days?: number;
  description?: string;
  rate_limit?: number;
}

/**
 * 创建 API Key 响应（含完整 key）
 */
export interface CreateApiKeyResponse extends ApiKey {
  key: string; // 完整的 API Key，仅返回一次
}

/**
 * 更新 API Key 请求
 */
export interface UpdateApiKeyRequest {
  name?: string;
  status?: ApiKeyStatus;
  scopes?: string[];
  expires_at?: string | null;
}

/**
 * API Key 验证响应
 */
export interface VerifyApiKeyResponse {
  valid: boolean;
  api_key_id?: number;
  user_id?: number;
  scopes?: string[];
  key_type?: string;
}

/**
 * API Key 统计响应
 */
export interface ApiKeyStatsResponse {
  user_id: number;
  total_keys: number;
  by_status: Record<string, number>;
  by_type: Record<string, number>;
}

/**
 * API Key 查询参数
 */
export interface ApiKeysQuery {
  key_type?: ApiKeyType;
  status?: ApiKeyStatus;
}
